{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "341dfa44",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "75ee20c2",
   "metadata": {
    "lang": "en"
   },
   "outputs": [],
   "source": [
    "orthogroups_file = \"important_species_orthogroups_at_lest_7_species.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f3568e4",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "orthogroups _file = \"90percent_ 1 _1_ Lepidoptera _MoreThan50Species_ orthogroups.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c222d52c",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "orthogroups _file = \"90percent_ 1 _1_ lepidoptera _MoreThan50Species_ orthogroups.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d4514b2",
   "metadata": {
    "lang": "fr"
   },
   "source": [
    "orthogroups _file = \"90percent_ 1 _1_ lepidoptera _MoreThan50Species_ orthogroups.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1d7a4017",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 存储每个物种的转录本 ID，字典键为 species（例如 Bombyx_mori），值为 set 类型的 ID 集合\n",
    "species_to_transcript_ids = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c6fa360b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 读取正交组文件（假设文件为 tab 分隔）\n",
    "with open(orthogroups_file, 'r') as infile:\n",
    "    reader = csv.DictReader(infile, delimiter='\\t')\n",
    "    # 遍历每一行\n",
    "    for row in reader:\n",
    "        # 遍历每个字段（第一列为 Orthogroup，其余为各物种）\n",
    "        for key, value in row.items():\n",
    "            if key == \"Orthogroup\":\n",
    "                continue  # 跳过第一列\n",
    "            # 如果对应单元格为空，则跳过\n",
    "            if not value.strip():\n",
    "                continue\n",
    "\n",
    "            # 物种名：将列名中的 \".protein\" 去掉\n",
    "            species = key.replace(\".protein\", \"\")\n",
    "            if species not in species_to_transcript_ids:\n",
    "                species_to_transcript_ids[species] = set()\n",
    "\n",
    "            # 如果单元格中有多个转录本 ID，以逗号分隔（并去掉多余空格）\n",
    "            ids = [tid.strip() for tid in value.split(\",\")]\n",
    "            for tid in ids:\n",
    "                species_to_transcript_ids[species].add(tid)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c0fc4e6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 用于从 GTF 属性中提取 transcript_id 的正则表达式\n",
    "transcript_id_re = re.compile(r'transcript_id \"([^\"]+)\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "434a2059",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "物种 Acyrthosiphon_pisum 的 exon 行已提取到 Acyrthosiphon_pisum.exon.gtf\n",
      "物种 Aedes_aegypti 的 exon 行已提取到 Aedes_aegypti.exon.gtf\n",
      "物种 Apis_mellifera 的 exon 行已提取到 Apis_mellifera.exon.gtf\n",
      "物种 Blattella_germanica 的 exon 行已提取到 Blattella_germanica.exon.gtf\n",
      "物种 Drosophila_mojavensis 的 exon 行已提取到 Drosophila_mojavensis.exon.gtf\n",
      "物种 Gryllus_bimaculatus 的 exon 行已提取到 Gryllus_bimaculatus.exon.gtf\n",
      "物种 Helicoverpa_armigera 的 exon 行已提取到 Helicoverpa_armigera.exon.gtf\n",
      "物种 Tribolium_castaneum 的 exon 行已提取到 Tribolium_castaneum.exon.gtf\n",
      "物种 Bombyx_mori 的 exon 行已提取到 Bombyx_mori.exon.gtf\n"
     ]
    }
   ],
   "source": [
    "## 遍历每个物种，处理对应的 GTF 文件\n",
    "for species, tid_set in species_to_transcript_ids.items():\n",
    "    # 构造 GTF 文件路径：假设在 /data/home/jinshuo/transcript/Genome/ 下，文件名为 {species}.gtf\n",
    "    gtf_file = f\"/data/home/jinshuo/transcript/Genome/{species}.gtf\"\n",
    "    if not os.path.exists(gtf_file):\n",
    "        print(f\"警告：文件 {gtf_file} 不存在，跳过 {species}！\")\n",
    "        continue\n",
    "\n",
    "    # 输出文件（你可以根据需要修改输出路径或文件名）\n",
    "    output_file = f\"{species}.exon.gtf\"\n",
    "    with open(gtf_file, 'r') as gf, open(output_file, 'w') as out_f:\n",
    "        for line in gf:\n",
    "            # 跳过 GTF 文件的注释行\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            # 确保至少有 9 列\n",
    "            if len(fields) < 9:\n",
    "                continue\n",
    "            feature = fields[2]\n",
    "            # 只处理 exon 行\n",
    "            if feature != \"exon\":\n",
    "                continue\n",
    "            attributes = fields[8]\n",
    "            # 提取 transcript_id\n",
    "            m = transcript_id_re.search(attributes)\n",
    "            if not m:\n",
    "                continue\n",
    "            transcript_id = m.group(1)\n",
    "            # 如果当前行的 transcript_id 存在于正交组中记录的该物种 ID 集合中，则输出此行\n",
    "            if transcript_id in tid_set:\n",
    "                out_f.write(line)\n",
    "    print(f\"物种 {species} 的 exon 行已提取到 {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "69b6fd47",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "df230c24",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bicyclus_anynana.exon.bed\n",
      "Bombyx_mori_exons.bed\n",
      "Heliconius_melpomene.exon.bed\n",
      "Plutella_xylostella.exon.fasta\n",
      "extract_orthogroups.ipynb\n",
      "pairwise.maf\n",
      "Bombyx_mori.exon.fasta\n",
      "extact_data.ipynb\n",
      "Pieris_napi.exon.fasta\n",
      "Blattella_germanica.exon.gtf\n",
      "Locusta_migratoria.exon.gtf\n",
      "Bombyx_mori.exon.gtf\n",
      "Junonia_coenia.exon.fasta\n",
      "Pieris_rapae.exon.gtf\n",
      "90percent_1_1_Lepidoptera_morethan100species_orthogroups.txt\n",
      "Heliconius_cydno.exon.bed\n",
      "exon_.bed_feature-1.fa\n",
      "nohup.out\n",
      "Bombyx_mori.genome.fa\n",
      "Spodoptera_litura.exon.gtf\n",
      "Helicoverpa_armigera.exon.gtf\n",
      "Lepidoptera_Orthogroups.tsv\n",
      "scriptsss.py\n",
      "Spodoptera_frugiperda.exon.bed\n",
      "Orthogroup\n",
      "out_sorted.maf\n",
      "Vanessa_cardui.exon.fasta\n",
      "Pieris_napi.exon.bed\n",
      "Lepidoptera_morethan90_species_list.txt\n",
      "script.py\n",
      "Lepidoptera_morethan100_species_list.txt\n",
      "output_region.maf\n",
      "new_intron_Bombyx_mori_exons.bed\n",
      "maf_sort.pl\n",
      "exec_mafft_alignment.ipynb\n",
      "Bombyx_mori.exon.bed\n",
      "Plutella_xylostella.exon.bed\n",
      "tissue\n",
      "Lepidoptera_species_list.txt\n",
      "Drosophila_ananassae.exon.gtf\n",
      "morethan30_species_list.txt\n",
      "Pieris_rapae.exon.bed\n",
      "test.log\n",
      "all_orthofinder_species_list.txt\n",
      "Vanessa_cardui.exon.bed\n",
      "Spodoptera_litura.exon.fasta\n",
      "Heliconius_melpomene.exon.fasta\n",
      "subset_tree.nk\n",
      "SpeciesTree_rooted.txt\n",
      "subset_tree.nwk\n",
      "Lepidoptera_morethan50_species_list.txt\n",
      "test_output_REV.mod\n",
      "Orthogroups.tsv\n",
      "Spodoptera_frugiperda.exon.fasta\n",
      "species.exon.bed\n",
      "scripts.py\n",
      "tmp.maf\n",
      "Heliconius_cydno.exon.bedfor\n",
      "Bombyx_mori_exons.sorted.bed\n",
      "test_nonAnc.maf\n",
      "Vanessa_cardui.exon.gtf.aln.fasta\n",
      "Acyrthosiphon_pisum.exon.gtf\n",
      "Bombyx_mori_as.bed\n",
      "all_species_list.txt\n",
      "90percent_1_1_Lepidoptera_morethan90species_orthogroups.txt\n",
      "90percent_1_1_Lepidoptera_morethan50species_orthogroups.txt\n",
      "test_output_model.mod\n",
      "Manduca_sexta.exon.bed\n",
      "Lepidoptera_morethan30_species_list.txt\n",
      "Junonia_coenia.exon.gtf\n",
      "og\n",
      "test.hal\n",
      "test_output_model_1.mod\n",
      "test-exon-1.maf\n",
      "test_nonAnc_origin.maf\n",
      ".ipynb_checkpoints\n",
      "Heliconius_cydno.exon.fasta\n",
      "Tribolium_castaneum.exon.gtf\n",
      "Aedes_aegypti.exon.gtf\n",
      "Pieris_rapae.exon.fasta\n",
      "Ostrinia_furnacalis.exon.gtf\n",
      "Ostrinia_furnacalis.exon.fasta\n",
      "test_exon1.maf\n",
      "test_nonAnc_sorted.maf\n",
      "Manduca_sexta.exon.fasta\n",
      "Bombyx_mori_as_output.csv\n",
      "Bicyclus_anynana.exon.fasta\n",
      "Helicoverpa_armigera.exon.bed\n",
      "output.maf\n",
      "Spodoptera_litura.exon.bed\n",
      "test_output_JC69.mod\n",
      "Bombyx_mori-Bicyclus_anynana.maf\n",
      "subset_tree_clean.nk\n",
      "test-Fat_body.maf\n",
      "Heliconius_erato.exon.bed\n",
      "mafft.sh\n",
      "output.aln\n",
      "test.ss\n",
      "Plutella_xylostella.exon.gtf\n",
      "test.wig\n",
      "important_species_list.txt\n",
      "Apis_mellifera.exon.gtf\n",
      "bombyx_mori_devas_new.csv\n",
      "Vanessa_cardui.exon.gtf\n",
      "Spodoptera_frugiperda.exon.gtf\n",
      "Bombyx_mori.genome.fa.fai\n",
      "Manduca_sexta.exon.gtf\n",
      "test_nonAnc_29.maf\n",
      "Heliconius_melpomene.exon.gtf\n",
      "Danaus_plexippus.exon.gtf\n",
      "mafft.log\n",
      "Helicoverpa_armigera.exon.fasta\n",
      "Pieris_napi.exon.gtf\n",
      "test_output_SSREV.mod\n",
      "exon_.background-0.fa\n",
      "testtest_nonAnc.ss\n",
      "test\n",
      "subset_tree_2.nwk\n",
      "test_output_HKY85+Gap.mod\n",
      "extract_devas_exon_to_bed.ipynb\n",
      "Junonia_coenia.exon.bed\n",
      "Heliconius_erato.exon.gtf\n",
      "Ostrinia_furnacalis.exon.bed\n",
      "important_species_orthogroups.txt\n",
      "test.mod\n",
      "Bicyclus_anynana.exon.gtf\n",
      "test.maf\n",
      "Danaus_plexippus.exon.bed\n",
      "Heliconius_erato.exon.fasta\n",
      "Danaus_plexippus.exon.fasta\n",
      "prepare_mafft.ipynb\n",
      "cactus_jobStore\n",
      "extract_genome_annoation.ipynb\n",
      "scriptss.py\n",
      "Heliconius_cydno.exon.gtf\n"
     ]
    }
   ],
   "source": [
    "for i in os.listdir():\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e72be0d4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base]",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  },
  "nbTranslate": {
   "displayLangs": [
    "*"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "en",
   "targetLang": "fr",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "186.4px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
